domains/client/relayer/src/worker.rs

```diff
@@ -9,8 +10,137 @@ use sp_messenger::messages::ChainId;
+    let mut chain_block_imported = client.every_import_notification_stream();
+    while let Some(imported_block) = chain_block_imported.next().await {
+        // if the client is in major sync, wait until sync is complete
+        if sync_oracle.is_major_syncing() {
+            tracing::debug!(target: LOG_TARGET, "Client is in major sync. Skipping...");
+            continue;
+        }
+
+        if !imported_block.is_new_best {
+            tracing::debug!(target: LOG_TARGET, "Imported non-best block. Skipping...");
+            continue;
+        }
+
+        let (block_hash, block_number) = match chain_id {
+            ChainId::Consensus => (
+                imported_block.header.hash(),
+                *imported_block.header.number(),
+            ),
+            ChainId::Domain(_) => {
+                // for domains, we gossip channel updates of imported block - 1
+                // since the execution receipt of the imported block is not registered on consensus
+                // without the execution receipt, we would not be able to verify the storage proof
+                let number = match imported_block.header.number().checked_sub(&One::one()) {
+                    None => continue,
+                    Some(number) => number,
+                };
+
+                let hash = match client.hash(number).ok().flatten() {
+                    Some(hash) => hash,
+                    None => {
+                        tracing::debug!(target: LOG_TARGET, "Missing block hash for number: {:?}", number);
+                        continue;
+                    }
+                };
+
+                (hash, number)
+            }
+        };
+
+        if let Err(err) = do_gossip_channel_updates::<_, _, CBlock>(
+            chain_id,
+            &client,
+            &gossip_message_sink,
+            block_number,
+            block_hash,
+        ) {
+            tracing::error!(target: LOG_TARGET, ?err, "failed to gossip channel update");
+        }
+    }
+}
+
+fn do_gossip_channel_updates<Client, Block, CBlock>(
+    src_chain_id: ChainId,
+    client: &Arc<Client>,
+    gossip_message_sink: &GossipMessageSink,
+    block_number: NumberFor<Block>,
+    block_hash: Block::Hash,
+) -> Result<(), Error>
+where
+    Block: BlockT,
+    CBlock: BlockT,
+    Client: BlockchainEvents<Block>
+        + HeaderBackend<Block>
+        + AuxStore
+        + ProofProvider<Block>
+        + ProvideRuntimeApi<Block>,
+    Client::Api: RelayerApi<Block, NumberFor<Block>, NumberFor<CBlock>, CBlock::Hash>,
+{
+    let api = client.runtime_api();
+    let api_version = api
+        .api_version::<dyn RelayerApi<Block, NumberFor<Block>, NumberFor<CBlock>, CBlock::Hash>>(
+            block_hash,
+        )?
+        .ok_or(sp_api::ApiError::Application(
+            "Failed to get relayer api version".into(),
+        ))?;
+
+    if api_version < 2 {
+        return Ok(());
+    }
+
+    let updated_channels = api.updated_channels(block_hash)?;
```

Comment:
Since the channel update storage is cleared on block initialization, the node only tries to send the `ChannelUpdate` message once and this message can also dropped during propagation then the issue still exists.

Reply:
Well, all operators will send this broadcast. If we keep the broadcasting, we will create another inefficiency of spamming network with unnecesssary channel updates.

The issue is never fully fixed where in if there is modified version of relayer, can choose to relay messages continuously and getting banned through txpool.

This change is to ensure honest operators do not get banned and they when to stop broadcasting.

Reply:
>The issue is never fully fixed where in if there is modified version of relayer, can choose to relay messages continuously and getting banned through txpool.
>This change is to ensure honest operators do not get banned and they when to stop broadcasting.

The issue we want to fix here is to prevent the honest operator from actively spamming the network by default, my concern is this can still happen if the `ChannelUpdate` message is dropped during propagation, which can happen as I observed in the devnet that XDM never success without the retrying mechanism.

>Well, all operators will send this broadcast. If we keep the broadcasting, we will create another inefficiency of spamming network with unnecesssary channel updates.

We can use some better strategies like:
- Low frequency, e.g. broadcast this message once per 100 blocks
- Only broadcast this message (with low frequency) if keep receiving already executed XDM

Reply:
> The issue we want to fix here is to prevent the honest operator from actively spamming the network by default, my concern is this can still happen if the ChannelUpdate message is dropped during propagation, which can happen as I observed in the devnet that XDM never success without the retrying mechanism.

Its a broadcast @NingLin-P. I believe you are confused regarding this. The issue you are seeing on devnet with out these changes where domain drop is expected since Domains might lag and state is not updated yet and txn will be dropped. Retry is the only way to get that message into tx pool and when the channel is updated, a broadcast is done automatically to signal src_chain operator to stop relaying this message. This is successfully verified on test.

Per my understanding, as long as the operators are connected to the network, they will receive the message. Only way the operators will drop this message from broadcast is src_chain operator is dropping due to failed conditions

The strategies you have suggested above will have the same issue. Likely

> Low frequency, e.g. broadcast this message once per 100 blocks

What if, per your understanding, this is dropped?
Reason I did not chose this frequential broadcast is avoid too many gossip messages for all the channels on a given domain. Broadcasting message when a specific channel is updated would be a spammy less broadcast IMHO

> Only broadcast this message (with low frequency) if keep receiving already executed XDM

The problem here is if the multiple reason
- There is already an existing txn in the tx-pool but state is not updated yet. So sending channel update is useless here since it will send an outdated channel state
- Domain is in the process of importing the domain block. Same issue as above regading gossiping the old channel state. Once the domain block is imported, their operators will send a broadcast anyway, so it there is no overall benefit here.


I would first prefer us to test this on devnet before looking for other strategies since IMHO, this change should e more than enough to signal operator to relay message since message is imported and executed

Reply:
Honestly, I'm not quite following about the difference between XDM and the channel state message, and the XDM is only relayed after the consensus block is finalized so I'm not sure the previous failure is related to lagging (which may require K blocks of lagging).

I'm also not following the "The problem here is if the multiple reason..." part, but yes I don't want to block this let's move on, plz test it on devnet to see how it goes though.